x-pack/legacy/plugins/infra/public/pages/logs/analysis/sections/anomalies/table.tsx

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useMemo, useState, useCallback } from 'react';
+import { EuiBasicTable, EuiButtonIcon } from '@elastic/eui';
+import { RIGHT_ALIGNMENT } from '@elastic/eui/lib/services';
+import { TimeRange } from '../../../../../../common/http_api/shared/time_range';
+import { GetLogEntryRateSuccessResponsePayload } from '../../../../../../common/http_api/log_analysis/results/log_entry_rate';
+import { AnomaliesTableExpandedRow } from './expanded_row';
+import { getTopAnomalyScoresByPartition } from '../helpers/data_formatters';
+
+interface TableItem {
+  id: string;
+  partition: string;
+  topAnomalyScore: number;
+}
+
+export const AnomaliesTable: React.FunctionComponent<{
+  results: GetLogEntryRateSuccessResponsePayload['data'];
+  setTimeRange: (timeRange: TimeRange) => void;
+  timeRange: TimeRange;
+}> = ({ results, timeRange, setTimeRange }) => {
+  const tableItems: TableItem[] = useMemo(() => {
+    return Object.entries(getTopAnomalyScoresByPartition(results)).map(([key, value]) => {
+      return {
+        id: key || 'unknown', // Note: EUI's table expanded rows won't work with a key of '' in itemIdToExpandedRowMap
+        partition: key || 'unknown',
+        topAnomalyScore: parseInt(Number(value).toFixed(0), 10),
+      };
+    });
+  }, [results]);
+
+  const [itemIdToExpandedRowMap, setItemIdToExpandedRowMap] = useState<
+    Record<string, React.ReactNode>
+  >({});
+
+  interface SortingOptions {
+    sort: {
+      field: string;
+      direction: string;
+    };
+  }
+
+  const [sorting, setSorting] = useState<SortingOptions>({
+    sort: {
+      field: 'topAnomalyScore',
+      direction: 'desc',
+    },
+  });
+
+  const handleTableChange = useCallback(
+    ({ sort = {} }) => {
+      const { field, direction } = sort;
+      setSorting({
+        sort: {
+          field,
+          direction,
+        },
+      });
+    },
+    [setSorting]
+  );
+
+  const sortedTableItems = useMemo(() => {
+    let sortedItems: TableItem[] = [];
+    if (sorting.sort.field === 'partition') {
+      sortedItems = tableItems.sort((a, b) => (a.partition > b.partition ? 1 : -1));
+    } else if (sorting.sort.field === 'topAnomalyScore') {
+      sortedItems = tableItems.sort((a, b) => a.topAnomalyScore - b.topAnomalyScore);
+    }
+    return sorting.sort.direction === 'asc' ? sortedItems : sortedItems.reverse();
+  }, [tableItems, sorting]);
+
+  const toggleExpandedItems = useCallback(
+    item => {
+      if (itemIdToExpandedRowMap[item.id]) {
+        const newItemIdToExpandedRowMap = {
+          ...itemIdToExpandedRowMap,
+        };
+        delete newItemIdToExpandedRowMap[item.id];
+        setItemIdToExpandedRowMap(newItemIdToExpandedRowMap);
```

Comment:
In the spirit of immuability, this is another possible way to express this:

```typescript
        const { [item.id]: toggledItem, ...remainingExpandedRowMap } = itemIdToExpandedRowMap;
        setItemIdToExpandedRowMap(remainingExpandedRowMap);
```

(also `delete`s often force deoptimization in JavaScript engines :wink:)